addons/test_mail_full/tests/test_mass_mailing.py

```diff
@@ -78,3 +106,109 @@ def test_mailing_w_blacklist_opt_out(self):
+    def test_mass_mailing_server_choice(self):
+        """Test that the right mail server is chosen to send the mailing.
+
+        Test also the envelop and the SMTP headers.
+        """
+        IrMailServer = type(self.env['ir.mail_server'])
+        find_mail_server = self.env['ir.mail_server']._find_mail_server
+
+        # Sanity check
+        self.assertEqual(self.env['ir.mail_server'].search_count([]), 4)
+
+        # Send one mailing
+        mailing = self.env['mailing.mailing'].create({
+            'subject': 'Mailing',
+            'email_from': 'std@odoo.com',
+        })
+        with patch.object(TestingSMTPSession, 'send_email') as send_email, \
+             patch.object(IrMailServer, '_find_mail_server', side_effect=find_mail_server) as patched_find_mail_server:
+            mailing.action_send_mail()
+        self.assertEqual(patched_find_mail_server.call_count, 1, 'Must be called only once')
+        send_email.assert_called_with(
+            smtp_from='std@odoo.com',
+            smtp_to_list=ANY,
+            message_from='std@odoo.com',
+            from_filter='std@odoo.com',
+        )
+
+        # Test sending mailing in batch
+        mailings = self.env['mailing.mailing'].create([{
+            'subject': 'Mailing',
+            'email_from': 'std@odoo.com',
+        }, {
+            'subject': 'Mailing',
+            'email_from': 'sqli@odoo.com',
+        }])
+        with patch.object(TestingSMTPSession, 'send_email') as send_email, \
+             patch.object(IrMailServer, '_find_mail_server', side_effect=find_mail_server) as patched_find_mail_server:
+            mailings.action_send_mail()
+        self.assertEqual(patched_find_mail_server.call_count, 2, 'Must be called only once per mail from')
+        send_email.assert_has_calls(
+            calls=[
+                call(
+                    smtp_from='std@odoo.com',
+                    smtp_to_list=ANY,
+                    message_from='std@odoo.com',
+                    from_filter='std@odoo.com',
+                ), call(
+                    # Must use the bounce address here because the mail server
+                    # is configured for the entire domain "odoo.com"
+                    smtp_from=MockValidator(lambda x: 'bounce' in x),
+                    smtp_to_list=ANY,
+                    message_from='sqli@odoo.com',
+                    from_filter='odoo.com',
+                ),
+            ],
+            any_order=True,
+        )
+
+        # We force a mail server on one mailing
+        mailings = self.env['mailing.mailing'].create([{
+            'subject': 'Mailing',
+            'email_from': 'std@odoo.com',
+        }, {
+            'subject': 'Mailing',
+            'email_from': 'test_force_mail_server@gmail.com',
+            'mail_server_id': self.env['ir.mail_server'].search([('from_filter', '=', 'notifications@odoo.com')]).id,
+        }])
+        with patch.object(TestingSMTPSession, 'send_email') as send_email, \
```

Comment:
Same remark as above: to be cleaned / sorted out as standard mock_mail tools and asserts.

Reply:
Used my own mock method